zetaclient/chains/evm/observer/inbound.go

```diff
@@ -98,83 +99,83 @@ func (ob *Observer) ProcessInboundTrackers(ctx context.Context) error {
 
 // ObserveInbound observes the evm chain for inbounds and posts votes to zetacore
 func (ob *Observer) ObserveInbound(ctx context.Context) error {
-	// get and update latest block height
-	blockNumber, err := ob.evmClient.BlockNumber(ctx)
-	switch {
-	case err != nil:
-		return errors.Wrap(err, "error getting block number")
-	case blockNumber < ob.LastBlock():
-		return fmt.Errorf("block number should not decrease: current %d last %d", blockNumber, ob.LastBlock())
-	}
-
-	ob.WithLastBlock(blockNumber)
+	logger := ob.Logger().Inbound.With().Str(logs.FieldMethod, "ObserveInbound").Logger()
 
-	// increment prom counter
-	metrics.GetBlockByNumberPerChain.WithLabelValues(ob.Chain().Name).Inc()
+	// keep last block up-to-date
+	if err := ob.updateLastBlock(ctx); err != nil {
+		return err
+	}
 
 	// uncomment this line to stop observing inbound and test observation with inbound trackers
 	// https://github.com/zeta-chain/node/blob/3879b5ef8b418542c82a4383263604222f0605c6/e2e/e2etests/test_inbound_trackers.go#L19
 	// TODO: implement a better way to disable inbound observation
 	// https://github.com/zeta-chain/node/issues/3186
 	//return nil
 
-	// skip if current height is too low
-	if blockNumber < ob.ChainParams().ConfirmationCount {
-		return fmt.Errorf("skipping observer, current block number %d is too low", blockNumber)
+	// calculate the unscanned block range
+	// Note: using separate scan range for each event incur more complexity (metrics, db, etc) and not worth it
+	startBlock, endBlock := ob.CalcUnscannedBlockRangeInboundSafe(config.BlockLimitPerScan)
+	if startBlock >= endBlock {
+		return nil
 	}
 
-	confirmedBlockNum := blockNumber - ob.ChainParams().ConfirmationCount
+	// observe inbounds in block range [startBlock, endBlock-1]
+	lastScannedNew := ob.observeInboundInBlockRange(ctx, startBlock, endBlock-1)
 
-	// skip if no new block is confirmed
-	lastScanned := ob.LastBlockScanned()
-	if lastScanned >= confirmedBlockNum {
-		return nil
+	// save last scanned block to both memory and db
+	if lastScannedNew > ob.LastBlockScanned() {
+		logger.Debug().Uint64("from", startBlock).Uint64("to", lastScannedNew).Msg("observed blocks for inbounds")
```

Comment:
Why if observeInboundInBlockRange already emit logs?

Reply:
This one is on `Debug` level and will not be printed in Datadog. It just allows us to debug locally and see the last scanned block without metrics, should we remove it?